code/Helper/Data.php

```diff
@@ -312,20 +304,24 @@ public function rebuildStoreProductIndex($storeId, $productIds)
         $emulationInfo = $this->startEmulation($storeId);
 
         try {
-            $collection = $this->product_helper->getProductCollectionQuery($storeId, $productIds);
-
+            $collection = $this->product_helper->getProductCollectionQuery($storeId, $productIds, false);
             $size = $collection->getSize();
 
+            if (!empty($productIds)) {
+                $size = max(count($productIds), $size);
+            }
```

Comment:
Because there might be IDs of deleted products which will never be in `$collection`